### PR TITLE
Pylexibank2

### PR DESCRIPTION
**Pull request by tresoldi**
This dataset is not ready yet, but it is worth discussing an important question on a PR. Please consider only the `_read_page_data()` for the time being (i.e., not `cmd_makecldf`), I still haven't started a FormSpec, and data is being written directly just to make it pass (I'll start working on that now).

The `satterthwaitetb` raw data has pretty much all problems one could expect. Along with normal form cleaning, in order not to modify the raw data we need to very carefully handle the results of the PDF conversion. I remember running into all sort of problems when I first did this, such as some page numbers being converted and others not, a couple of Unicode issues and, worst of all, cases where the order in the data is wrong (i.e., the published list has "Lang_a, Lang_b, Lang_c", but in the conversion we ended with "Lang_b, Lang_a, Lang_c"). The only way to find it out was to manually check the results.

I have separated the logic for fixing all such problems in a `_read_page_data()` method, along with my three replacement data structures currently as global ones (`ERRATA`, `DROP_FINAL`, and `ROW_IDX`, at the top). Most important, I heavily commented the code, explaining what and why I am doing, and moving to dumb logic whenever possible (such as replacing list comprehensions with nested `for` loops).

The question is: is this a good way to handle such cases? The idea is that `_read_page_data()` returns a list of dictionary as it would if I were allowed to manually edit the raw data. What else should I change? The global data structures (like `ERRATA`) could go to an `etc/` file, if desired.

### CODE REVIEW

File: lexibank_satterthwaitetb.py [hunk 1]
```diff
@@ -1,192 +1,261 @@
...
 import itertools
+from pathlib import Path
 import re
 
 from clldutils.misc import slug
-from clldutils.path import Path
-from clldutils.text import split_text, strip_brackets
 from pylexibank.dataset import Dataset as BaseDataset
-from pylexibank.util import getEvoBibAsBibtex, pb
+from pylexibank.util import progressbar
```
Comment (LinguList) [inline lexibank_satterthwaitetb.py]:
```python
from pylexibank import progressbar
```
@xrotwang emphasized this, we should import in this way

File: lexibank_satterthwaitetb.py [hunk 2]
```diff
@@ -1,192 +1,261 @@
...
-                        if lang in ["concepts", "Hanzi"]:
+    def cmd_makecldf(self, args):
+        # read page data, solving all PDF conversion problems (forms still
+        # need to be cleaned)
+        page_data = self._read_page_data()
+
+        # write the bibliographic sources
+        args.writer.add_sources()
+
+        # add the languages from the language list
```
Comment (LinguList) [inline lexibank_satterthwaitetb.py]: Please don't make ids by slugging, but maybe you're not there yet, right? But There should be a list with languages, CamelCase, in `etc/languages.tsv`.

File: lexibank_satterthwaitetb.py [hunk 3]
```diff
@@ -1,192 +1,261 @@
...
+            args.writer.add_language(
+                ID=slug(language["Name"]),
+                Name=language["Name"],
+                Glottocode=language["Glottocode"],
+                Glottolog_Name=language["Glottolog_Name"],
+            )
+
+        # add the concepts from the concept list
+        # TODO: add `English` and other fields?
+        for concept in self.conceptlist.concepts.values():
```
Comment (LinguList) [inline lexibank_satterthwaitetb.py]: Concept-class ideally follows what the concepticon list gives. So you may need a CustomConcept

File: lexibank_satterthwaitetb.py [hunk 4]
```diff
@@ -1,192 +1,261 @@
...
+                for lang in page.keys():
+                    if lang in ["concepts", "Hanzi"]:
+                        continue
+                    # extract the value (raw transcription) and apply
+                    # some initial correction before splitting into the
+                    # various forms; before removing parentheses, we
+                    # manually correct and preserve the few cases in which
+                    # the parentheses are actual sound information and
+                    # not comments
+                    value = page[lang][idx]
```
Comment (LinguList) [inline lexibank_satterthwaitetb.py]: of course, for the formspec.

**Comment by tresoldi** [comment]
@LinguList please see my PR comment: https://github.com/lexibank/satterthwaitetb/pull/2#issue-334204527 -- I've already started working on the normal things like FormSpec, I'm asking for your opinion on the approach towards the raw data (i.e., the `_read_page_data()` method).

**Comment by tresoldi** [comment]
The last commit improves a number of things, but the dataset is not reading for merging. There are orthographic profile issues to solve, but those need to wait for decision on FormSpec (depending on https://github.com/lexibank/diacl/issues/9 , https://github.com/lexibank/pylexibank/issues/145 , https://github.com/lexibank/pylexibank/issues/142 , and https://github.com/lexibank/pylexibank/issues/146 )

**Comment by LinguList** [comment]
Yes, I understood this, but I did not understand, why you would not just simply use the data in the excel spreadsheet that Damian sent to me years ago. I mean, if it is that complicated, why go for it? I have worked with other datasets, and whenever there's something manual to do, it is better to convert the data into some intermediate format, and THEN manually correct it, etc. 

Even if this is a bit contra the idea of replicability: we are heavily retro-standardizing data here, and some datasets nobody really ever will care about, apart from us for a purpose like CLICS, so why go through the pain for a dataset that is not even segmentizable? 

I'd say: follow Sunzi's Art of War: Know when you should fight and know when you should not fight. This is a case, where I think it is not worth to fight a lot.

**Review by LinguList — COMMENTED** [review]
Note that we explicitly say that we do not want to have this in segmented form. The data mixes different orthographies, it is only possible to do this langauge-by-language, but we do not have time for this, and it is not clear if it actually worth it.

**Comment by tresoldi** [comment]
No segmentation, then? Should I delete the profile as well, or just keep it around in case we go for a per-language segmentation in the future?

**Comment by LinguList** [comment]
> No segmentation, then? Should I delete the profile as well, or just keep it around in case we go for a per-language segmentation in the future?

Yes, as I said [here](https://github.com/lexibank/pylexibank/issues/142#issuecomment-547841634): the data is in my opinion not segmentable.

**Comment by LinguList** [comment]
And you can also see that we never had a segmented version, just a `profile.tsv` which should be deleted.

**Comment by tresoldi** [comment]
Following the discussion, I've removed the segmentation and generated a new source file. Considering that the one in there had already been generated by us, using `pdf2txt`, any restriction about not modifying the source stops applying (but I am still keeping the output of `pdf2txt` and the code used to fix it, so there is full transparency).

**Comment by tresoldi** [comment]
Will get back to this PR once we decide about the FormSpec issues, but I'd say it is pretty much done.

**Comment by LinguList** [comment]
I think, once @chrzyki does work on IDS and adds a tiny formspec modification, and we have the list-of-tuples or so in there, we can just copy-paste this. OR we add it even now to pylexibank. But you'd have to indicate a number >= 2.0.1 or similar for compatibility...

**Comment by tresoldi** [comment]
Updated with the new FormSpec. Please check, @LinguList , @SimonGreenhill , @chrzyki , @xrotwang

**Review by SimonGreenhill — APPROVED** [review]

**Review by chrzyki — APPROVED** [review]

**Review by LinguList — CHANGES_REQUESTED** [review]
Could you please quickly correct the identifiers? I thought I mentioned this before, but we have a quasi-format for concept-ids, and we decided to not slug lowercase in languages. Please, this is important as it also helps avoiding problems, and it enhances readability...

**Comment by tresoldi** [comment]
@LinguList : but what should I use? The labels cannot be used directly as CLDF identifiers (for example, `be born, to `)

**Comment by LinguList** [comment]
Wasn't that clear from our discussion? I mean, I used it in all my 14 datasets:

```python
concept_lookup = args.writer.add_concepts(
    id_factory=lambda x: x.split('-')[-1]+'_'+slug(x.english),
    lookup_factory='Name')
```

That's all it needs.

And for languages, if you just make the ID capitalized, all you'll need is:

```python
language_lookup = args.writer.add_languages(lookup_factory="Name")
```

**Comment by LinguList** [comment]
Check here, for what I just did in `bowernpny`: https://github.com/lexibank/bowernpny/blob/abebb32e81a13aa295a01652d0b012c53ad89bd7/lexibank_bowernpny.py#L81-L86

**Comment by tresoldi** [comment]
@LinguList I suppose now it is clear: you want to manually slug the concept identifiers, but not the language ones. I'll fix it.

**Comment by tresoldi** [comment]
@LinguList , could you please check if it is correct now?

**Comment by LinguList** [comment]
Number is okay, but we decided on `x.id`, I thought you were present, when we discussed this, it was two weeks ago, in @chrzyki's room, when we decided to go camel case for language identifiers and using this construct with the always unique part of the concepticon id, which is not necessarily identical with the number, for the concept identifiers to increase transparency, which is especially important when checking `cldf/forms.csv`.

**Comment by tresoldi** [comment]
@LinguList I was present but understood this was something for when lexibank would be released, not for these CLICS datasets (perhaps reinforced by some datasets which don't follow these specs being merged).

There were two IDs to change to have in CamelCase, `LahuShi` and `SichuanYi`, they are fixed with the last commit. But could you point me to what I should have in place of the `number` (not only here, in general)? In detail, how should I use the unique part of the concepticon id in cases when there is no concepticon mapping?

**Comment by LinguList** [comment]
You use the ID for a concept provided by concepticon, not the concepticon id. 

```python
lambda x: x.id.split('-')[-1]+'_'+slug(x)
```

You don't use concepticon_id, that would be:

```python
lambda x: x.concepticon_id
```

**Comment by tresoldi** [comment]
Ok, sorry for all the questioning, but I was feeling really stupid about it. I had actually used that code before, it happens that in this dataset, by an unlucky coincidence, the resulting IDs are the same.

I've changed the code and made the cldf. Please confirm it is right, I'll copy this factory id to the other ones once it is right.

**Review by LinguList — APPROVED** [review]
Isn't it better like this? It is helpful to look at the other PRs, as I requested this from almost everybody in the past week. And of course it does not make sense to ignore it now and then re-do it for all of lexibank, although one could discuss defaulting to this behavior. And the good thing (I also explained that some place): if your id is only numeric, since it resides still in `etc/concepts`, so it does not  have the dashes, it will be the same result for the parameter-id.

**Pull request merged**